### ISSUE TITLE
More precise assessment of RDC cost

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#51]
On July 25 during the hackathon @oliviermattelaer removed CUDA Relocatable Device Code, ie "-rdc=true"
https://github.com/madgraph5/madgraph4gpu/commit/bc03049a84968d94f814a91905a787a9aa995af3
We found at the time that this was giving us a significant performance boost. It would be useful however yo go back there and assess more precisely how much this gains (less than 1%, some 10%, a factor 10?...).
The drawback of that choice was that we must use a single file compilation, with no loaded libraries. This is not a problem for the moment, but as we get into more complicated workflows (eg with MadEvent and PDFs) this may change, so it is important to quantify the benefit in having no rdc (or the cost in having back rdc).

**Comment by valassi** [comment]
As discussed now with @oliviermattelaer , probably at some point I got rid of FFVxxx functions from the headers, because it was faster. I mean, in cuda these must be in the same source anyway, so there is no need for a header. Then in the end I did not add it back for c++ either, as also in c++ these functions are now called within the same file where they are defined and only there (CPPProcess.cc). This may need to be reviewed, especially if it turns out that we can move away from single source and use RDC again

**Comment by oliviermattelaer** [comment]
In the hackathon, the assesment was that for 
e+ e- > mu+ mu- the gain was HUGE.
The slide from the hackathon was mentionning a 13x factor (but a reduce occupance by a factor of 2 due to higher number of register).
The code passed from being memory bound to compute bound 

For g g > t t~ g g
the speed-up was a factor of two (1.9x):
In this case, we face issue that we did not have enough register available and therefore we were still memory bound.

**Comment by oliviermattelaer** [comment]
[Hackaton_day2.pdf](https://github.com/madgraph5/madgraph4gpu/files/5567863/Hackaton_day2.pdf)

Here is the presentation of the hackaton with the associated details

**Comment by valassi** [comment]
> 
> 
> [Hackaton_day2.pdf](https://github.com/madgraph5/madgraph4gpu/files/5567863/Hackaton_day2.pdf)
> 
> Here is the presentation of the hackaton with the associated details

Wow, great, thanks a lot! I did not remember we had all this material still from the hackathon ;-)

**Comment by valassi** [comment]
Note the similarities of RDC in CUDA to LTO and inlining in C++ (issue #229 ... there factors 2-4 seem to be possible speedups)

**Comment by valassi** [comment]
I have done a few recent tests for this in PR #332.

There is certainly a cost for RDC, but with our latest code it is much less than the factor 10 we were discussing two years ago.

Just adding the "-rdc true" flag, without moving code around:
- eemumu is actually faster by 20%! (1.35 vs 1.11) but this is only because it uses two registers less somehow, going below a threshold from 130 to 128 registers - see https://github.com/madgraph5/madgraph4gpu/pull/332/commits/da08e56f69909a9f5836222cbdb2a44ae5c16071
- ggtt speed is essentially unchanged, but again fewer registers (170 vs 172) are used - see https://github.com/madgraph5/madgraph4gpu/pull/332/commits/7617ad09d6c1cbff2ee226f4e286bb04ff5c8dff
- ggttgg slowdown is around 1% at peak throughput (4.84 instead of 4.89) and 5% for lower gpublocks/gputhreads (4.17 vs 4.39) - see https://github.com/madgraph5/madgraph4gpu/pull/332/commits/5ef2bf248c7f1c66c91e265bdab66f7806731a7d

The slowdown in ggttgg is more important if HelAmps.cc is compiled as a separate object
- ggttgg slowdown is 15% (4.15 vs 4.85) at peak, see https://github.com/madgraph5/madgraph4gpu/pull/332/commits/3951a6e0d187d3071cf8ae41f05479edea62ea83
- note that compiling HelAmps.cc as a separate object is only possible if "-rdc true" is used, as "-rdc false" gives runtime failures/aborts - see https://github.com/madgraph5/madgraph4gpu/pull/332/commits/76a7749b0b19c34cc3744fe3ab9dcd464ab73d59
- note also that if -rdc is not specified, this is equivalent to "-rdc false", see https://github.com/madgraph5/madgraph4gpu/pull/332/commits/17d90a322022d08050c0ab371a3d67d7ce746758

I would conclude
- we keep things as they are, with no -rdc flag, which means "-rdc false"

I would say that this can be closed.

**Comment by valassi** [comment]
PS Just to say, these tests were done while I was working on #328 (apiwf), where I introduced templated FFV functions to prepare for wavefunctions/amplitudes in global mamory, which is a prerequisite to splitting kernels. In this patch I moved HelAmps to being fully inlines in the .h file.

The motivation for doing these RDC tests is that the move of FFV functions to template functions (whether with an explicit inline parameter or not) increased build times with cuda 11.1 enormously. It took 75 minutes to compile gCPPProcess.o. I wanted to understand whether compiling the FFVs in a separate HelAmps.o rather than withing gCPPProcess.o would be feasible and would speed up compilation. In the end I fixed that issue in a completely different (unexpected) way, that is moving back to cuda 11.5.
